src/adapters/supabase/helpers/log.ts

```diff
@@ -0,0 +1,196 @@
+import { getAdapters, getBotContext } from "../../../bindings";
+import { Payload } from "../../../types";
+import { getOrgAndRepoFromPath } from "../../../utils/private";
+interface Log {
+  repo: string | null;
+  org: string | null;
+  commentId: number | undefined;
+  issueNumber: number | undefined;
+  logMessage: string;
+  errorType: levels;
+  timestamp: string;
+}
+
+export enum levels {
+  error = 0,
+  warn = 1,
+  info = 2,
+  http = 3,
+  verbose = 4,
+  debug = 5,
+  silly = 6,
+}
+
+export class GitHubLogger {
+  private supabase = getAdapters().supabase;
+  private maxLevel;
+  private app;
+  private logEnvironment;
+  private logQueue: Log[] = []; // Your log queue
+  private maxConcurrency = 5; // Maximum concurrent requests
+  private retryLimit = 3; // Maximum number of retries
+  private retryDelay = 1000; // Delay between retries in milliseconds
+  private throttleCount = 0;
+
+  constructor(app: string, logEnvironment: string, maxLevel: keyof typeof levels) {
+    this.app = app;
+    this.logEnvironment = logEnvironment;
+    this.maxLevel = levels[maxLevel];
+  }
+
+  async sendLogsToSupabase({ repo, org, commentId, issueNumber, logMessage, errorType, timestamp }: Log) {
+    try {
+      const { error } = await this.supabase.from("log_entries").insert([
+        {
+          repo_name: repo,
+          level: errorType,
+          org_name: org,
+          comment_id: commentId,
+          log_message: logMessage,
+          issue_number: issueNumber,
+          timestamp,
+        },
+      ]);
+
+      if (error) {
+        console.error("Error logging to Supabase:", error.message);
+        return;
+      }
+    } catch (error) {
+      if (error instanceof Error) {
+        console.error("An error occurred:", error.message);
+      }
+    }
+  }
+
+  async processLogs(log: Log) {
+    try {
+      await this.sendLogsToSupabase(log);
+    } catch (error) {
+      console.error("Error sending log, retrying:", error);
+      await this.retryLog(log);
+    }
+  }
+
+  async retryLog(log: Log, retryCount = 0) {
+    if (retryCount >= this.retryLimit) {
+      console.error("Max retry limit reached for log:", log);
+      return;
+    }
+
+    await new Promise((resolve) => setTimeout(resolve, this.retryDelay));
+
+    try {
+      await this.sendLogsToSupabase(log);
+    } catch (error) {
+      console.error("Error sending log (after retry):", error);
+      await this.retryLog(log, retryCount + 1);
+    }
+  }
+
+  async processLogQueue() {
+    while (this.logQueue.length > 0) {
+      const log = this.logQueue.shift();
+      if (!log) {
+        continue;
+      }
+      await this.processLogs(log);
+    }
+  }
+
+  async throttle(fn: () => void) {
+    if (this.throttleCount >= this.maxConcurrency) {
```

Comment:
@devpanther So if there are 6 concurrent "save log" requests while `this. maxConcurrency === 5` then the 6th "save log" request is no saved to a DB, right?

Reply:
> @devpanther So if there are 6 concurrent "save log" requests while `this. maxConcurrency === 5` then the 6th "save log" request is not saved to a DB, right?

It remains in the queue and will be sent once the rest are resolved... so it does not overwhelm the supabase database

Reply:
http/s max concurrency by default in Chrome is 6. Perhaps that is the optimal amount. Why did you choose 5?

Reply:
> http/s max concurrency by default in Chrome is 6. Perhaps that is the optimal amount. Why did you choose 5?

Just a choice, but I made it 6 now